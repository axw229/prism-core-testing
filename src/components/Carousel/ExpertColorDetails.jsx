// @flow
import React from 'react'
import { fullColorNumber, getContrastYIQ } from '../../../src/shared/helpers/ColorUtils'
import type { Color } from '../../shared/types/Colors'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheckCircle } from '@fortawesome/pro-solid-svg-icons'
import { faPlusCircle } from '@fortawesome/pro-light-svg-icons'
import { connect } from 'react-redux'
import { add } from '../../store/actions/live-palette'
import some from 'lodash/some'
import './ExpertColorDetails.scss'

type Props = {
    expertColors: Color[],
    addColors: Color
}
const baseClass = 'prism-expert-color-details'
const icons = 'toggle-check-icons'

export function ExpertColorDetails (props: Props) {
  const { expertColors, addColors } = props
  const topSectionColor = expertColors[0]
  const bottomCollor = expertColors.slice(1)
  const renderColorWall = (color, props, key) => {
    const isColorAdded = some(addColors, color)
    return (
      <div key={key} className={`${baseClass}__content__wrapper`} style={{ backgroundColor: color.hex, color: getContrastYIQ(color.hex) }}>
        <div className={`${baseClass}__content__wrapper__color-number`} >
          {fullColorNumber(color.brandKey, color.colorNumber)}
        </div>
        <div className={`${baseClass}__content__wrapper__color-name`}>
          {color.name}
        </div>
        {
          isColorAdded &&
          <button onClick={() => handleClick(color, props)}>
            <FontAwesomeIcon style={{ color: getContrastYIQ(color.hex) }} className={`${baseClass}__content__wrapper__${icons}`} icon={faCheckCircle} />
          </button>
        }
        {
          !isColorAdded &&
            <button onClick={() => handleClick(color, props)}>
              <FontAwesomeIcon style={{ color: getContrastYIQ(color.hex) }} className={`${baseClass}__content__wrapper__${icons}`} icon={faPlusCircle} />
            </button>
        }
      </div>)
  }
  return (
    <div className={`${baseClass}`}>
      <div className={`${baseClass}__header`}>
        <div className={`${baseClass}__header__title`}>Seleted from Lifestyle - Chrysalls</div>
        <button className={`${baseClass}__header__link`}>
          VIEW FULL COLLECTION
        </button>
      </div>
      <div className={`${baseClass}__top__section`}>
        {
          renderColorWall(topSectionColor, props)
        }
      </div>
      <div className={`${baseClass}__bottom__section`}>
        {
          bottomCollor.map((color, key) => {
            return renderColorWall(color, props, key)
          })
        }
      </div>
    </div>
  )
}

const handleClick = (color: Color, props: Object) => {
  const { addToLivePalette } = props
  addToLivePalette(color)
}

const mapStateToProps = (state, props) => {
  const { lp } = state
  return {
    addColors: lp.colors
  }
}

const mapDispatchToProps = (dispatch: Function) => {
  return {
    addToLivePalette: (color) => {
      dispatch(add(color))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpertColorDetails)
