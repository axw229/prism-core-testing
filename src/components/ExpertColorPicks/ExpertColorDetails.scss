.prism-expert-color-details {
  display: flex;
  justify-content: center;
  flex-direction: column;
  &__header{
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 1em 0;
    &__title{
      text-align: center;
    }
    &__link {
      color: #2cabe2;
      font-size: 0.7rem;
      font-weight: 600;
      margin: 0.7em 0 0 0;
      outline: none;
    }
  }
  &__bottom__section {
    border-top: 1px solid #ffffff;
    display: flex;
  }
  &__top__section &__content__wrapper {
    padding-bottom: 200px;
  }
  &__bottom__section &__content__wrapper {
    padding-bottom: 100px;
  }
  &__content__wrapper {
    flex: 50% 0 0;
    position: relative;
    padding: 1em;
    &__toggle-check-icons {
      font-size: 2em;
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      &:focus {
        outline-style: solid;
        outline-color: #2cabe2;
        outline-width: 2px;
        z-index: 100;
      }
    }
    &__color-name {
      font-weight: 700;
    }
  }
}
