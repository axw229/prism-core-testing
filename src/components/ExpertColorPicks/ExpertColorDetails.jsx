// @flow
/**
 * expert color details component
 * loads when you click on the expert color swatch
 */

import React from 'react'
import {
  fullColorNumber,
  getContrastYIQ
} from '../../../src/shared/helpers/ColorUtils'
import type { Color } from '../../shared/types/Colors'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheckCircle } from '@fortawesome/pro-solid-svg-icons'
import { faPlusCircle } from '@fortawesome/pro-light-svg-icons'
import { connect } from 'react-redux'
import { add } from '../../store/actions/live-palette'
import some from 'lodash/some'
import './ExpertColorDetails.scss'

type Props = {
    expertColors: Color[],
    addColors: Color
}
const baseClass = 'prism-expert-color-details'
const icons = 'toggle-check-icons'

export function ExpertColorDetails (props: Props) {
  const { expertColors, addColors } = props
  const topSectionColor = expertColors.colorDefs[0]
  const bottomCollor = expertColors.colorDefs.slice(1)
  const renderColorWall = (color, props, key) => {
    const isColorAdded = some(addColors, color)
    return (
      <div
        className={`${baseClass}__content__wrapper`}
        key={key}
        style={{
          backgroundColor: color.hex,
          color: getContrastYIQ(color.hex)
        }}
      >
        <div className={`${baseClass}__content__wrapper__color-number`}>
          {fullColorNumber(color.brandKey, color.colorNumber)}
        </div>
        <div className={`${baseClass}__content__wrapper__color-name`}>
          {color.name}
        </div>
        {
          isColorAdded &&
          <button onClick={() => handleClick(color, props)}>
            <FontAwesomeIcon
              style={{ color: getContrastYIQ(color.hex) }}
              className={`${baseClass}__content__wrapper__${icons}`}
              icon={faCheckCircle}
            />
          </button>
        }
        {
          !isColorAdded &&
            <button onClick={() => handleClick(color, props)}>
              <FontAwesomeIcon
                style={{ color: getContrastYIQ(color.hex) }}
                className={`${baseClass}__content__wrapper__${icons}`}
                icon={faPlusCircle}
              />
            </button>
        }
      </div>)
  }
  return (
    <div className={`${baseClass}`}>
      <div className={`${baseClass}__header`}>
        <div className={`${baseClass}__header__title`}>
          Seleted from {expertColors.name}
        </div>
        {/*
            <button className={`${baseClass}__header__link`}>
                TODO:noah.hall
                requires each collection summary to have a URI
              VIEW FULL COLLECTION
            </button>
        */}
      </div>
      <div className={`${baseClass}__top__section`}>
        {
          renderColorWall(topSectionColor, props)
        }
      </div>
      <div className={`${baseClass}__bottom__section`}>
        {
          bottomCollor.map((color, key) => {
            return renderColorWall(color, props, key)
          })
        }
      </div>
    </div>
  )
}

const handleClick = (color: Color, props: Object) => {
  const { addToLivePalette } = props
  addToLivePalette(color)
}

const mapStateToProps = (state, props) => {
  const { lp } = state
  return {
    addColors: lp.colors
  }
}

const mapDispatchToProps = (dispatch: Function) => {
  return {
    addToLivePalette: (color) => {
      dispatch(add(color))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpertColorDetails)
