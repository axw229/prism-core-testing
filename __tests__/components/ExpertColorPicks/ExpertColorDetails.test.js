import React from 'react'
import { Provider } from 'react-redux'
import { shallow, mount } from 'enzyme'
import ExpertColorDetails from 'src/components/ExpertColorPicks/ExpertColorDetails'

const livePaletteColors = [{ hex: '#c4c4c4', brandKey: 'SW', colorNumber: 90212, name: 'Blue' }]
const dispatch = jest.fn(({ payload: { color } }) => livePaletteColors.push(color))

const getExpertColorDetails = () => {
  return mount(
    <Provider store={{
      getState: () => ({ lp: { colors: livePaletteColors } }),
      subscribe: jest.fn(),
      dispatch: dispatch
    }}>
      <ExpertColorDetails expertColors={{
        colorDefs: [
          { hex: '#cdd2d2', brandKey: 'SW', colorNumber: 90210, name: 'Bobafett' },
          { hex: '#a2a2a2', brandKey: 'SW', colorNumber: 90211, name: 'Green' },
          { hex: '#c4c4c4', brandKey: 'SW', colorNumber: 90212, name: 'Blue' }
        ]
      }} />
    </Provider>)
}

let expertColorDetails

beforeEach(() => expertColorDetails = getExpertColorDetails())

it('renders a color wall for each color', () => {
  const colorWalls = expertColorDetails.find('.prism-expert-color-details__content__wrapper')
  expect(colorWalls).toHaveLength(3)
  expect(colorWalls.get(0).props.style.backgroundColor).toBe('#cdd2d2')
  expect(colorWalls.get(1).props.style.backgroundColor).toBe('#a2a2a2')
  expect(colorWalls.get(2).props.style.backgroundColor).toBe('#c4c4c4')
})

it('shows a faCheckCircle for colors that are already present in the live-palette', () => {
  expect(expertColorDetails.find('svg[data-icon="check-circle"]')).toHaveLength(1)
  expect(expertColorDetails.find('svg[data-icon="plus-circle"]')).toHaveLength(2)
})

it('clicking a button calls add dispatch function and updates icons', () => {
  expertColorDetails.find('button').first().simulate('click')
  expect(dispatch.mock.calls[0][0]).toEqual({
    payload: {
      color: { brandKey: 'SW', colorNumber: 90210, hex: '#cdd2d2', name: 'Bobafett' }
    },
    type: 'ADD_LP_COLOR'
  })
})
