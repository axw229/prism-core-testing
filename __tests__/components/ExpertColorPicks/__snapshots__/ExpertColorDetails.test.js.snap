// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction],
      "getState": [Function],
      "subscribe": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
          Array [
            [Function],
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    }
  }
>
  <ExpertColorDetails
    expertColors={
      Object {
        "colorDefs": Array [
          Object {
            "brandKey": "SW",
            "colorNumber": 90210,
            "hex": "#cdd2d2",
            "name": "Bobafett",
          },
          Object {
            "brandKey": "SW",
            "colorNumber": 90211,
            "hex": "#a2a2a2",
            "name": "Green",
          },
          Object {
            "brandKey": "SW",
            "colorNumber": 90212,
            "hex": "#c4c4c4",
            "name": "Blue",
          },
        ],
      }
    }
  >
    <div
      className="prism-expert-color-details"
    >
      <div
        className="prism-expert-color-details__header"
      >
        <div
          className="prism-expert-color-details__header__title"
        >
          Seleted from 
        </div>
      </div>
      <div
        className="prism-expert-color-details__top__section"
      >
        <div
          className="prism-expert-color-details__content__wrapper"
          style={
            Object {
              "backgroundColor": "#cdd2d2",
              "color": "#000",
            }
          }
        >
          <div
            className="prism-expert-color-details__content__wrapper__color-number"
          >
            SW 90210 
          </div>
          <div
            className="prism-expert-color-details__content__wrapper__color-name"
          >
            Bobafett
          </div>
          <button
            onClick={[Function]}
          >
            <FontAwesomeIcon
              border={false}
              className="prism-expert-color-details__content__wrapper__toggle-check-icons"
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "fal",
                  "plus-circle",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              style={
                Object {
                  "color": "#000",
                }
              }
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-plus-circle fa-w-16 prism-expert-color-details__content__wrapper__toggle-check-icons"
                data-icon="plus-circle"
                data-prefix="fal"
                focusable="false"
                role="img"
                style={
                  Object {
                    "color": "#000",
                  }
                }
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M384 250v12c0 6.6-5.4 12-12 12h-98v98c0 6.6-5.4 12-12 12h-12c-6.6 0-12-5.4-12-12v-98h-98c-6.6 0-12-5.4-12-12v-12c0-6.6 5.4-12 12-12h98v-98c0-6.6 5.4-12 12-12h12c6.6 0 12 5.4 12 12v98h98c6.6 0 12 5.4 12 12zm120 6c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zm-32 0c0-119.9-97.3-216-216-216-119.9 0-216 97.3-216 216 0 119.9 97.3 216 216 216 119.9 0 216-97.3 216-216z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </button>
        </div>
      </div>
      <div
        className="prism-expert-color-details__bottom__section"
      >
        <div
          className="prism-expert-color-details__content__wrapper"
          key="0"
          style={
            Object {
              "backgroundColor": "#a2a2a2",
              "color": "#000",
            }
          }
        >
          <div
            className="prism-expert-color-details__content__wrapper__color-number"
          >
            SW 90211 
          </div>
          <div
            className="prism-expert-color-details__content__wrapper__color-name"
          >
            Green
          </div>
          <button
            onClick={[Function]}
          >
            <FontAwesomeIcon
              border={false}
              className="prism-expert-color-details__content__wrapper__toggle-check-icons"
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "fal",
                  "plus-circle",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              style={
                Object {
                  "color": "#000",
                }
              }
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-plus-circle fa-w-16 prism-expert-color-details__content__wrapper__toggle-check-icons"
                data-icon="plus-circle"
                data-prefix="fal"
                focusable="false"
                role="img"
                style={
                  Object {
                    "color": "#000",
                  }
                }
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M384 250v12c0 6.6-5.4 12-12 12h-98v98c0 6.6-5.4 12-12 12h-12c-6.6 0-12-5.4-12-12v-98h-98c-6.6 0-12-5.4-12-12v-12c0-6.6 5.4-12 12-12h98v-98c0-6.6 5.4-12 12-12h12c6.6 0 12 5.4 12 12v98h98c6.6 0 12 5.4 12 12zm120 6c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zm-32 0c0-119.9-97.3-216-216-216-119.9 0-216 97.3-216 216 0 119.9 97.3 216 216 216 119.9 0 216-97.3 216-216z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </button>
        </div>
        <div
          className="prism-expert-color-details__content__wrapper"
          key="1"
          style={
            Object {
              "backgroundColor": "#c4c4c4",
              "color": "#000",
            }
          }
        >
          <div
            className="prism-expert-color-details__content__wrapper__color-number"
          >
            SW 90212 
          </div>
          <div
            className="prism-expert-color-details__content__wrapper__color-name"
          >
            Blue
          </div>
          <button
            onClick={[Function]}
          >
            <FontAwesomeIcon
              border={false}
              className="prism-expert-color-details__content__wrapper__toggle-check-icons"
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "fa",
                  "check-circle",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              style={
                Object {
                  "color": "#000",
                }
              }
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-check-circle fa-w-16 prism-expert-color-details__content__wrapper__toggle-check-icons"
                data-icon="check-circle"
                data-prefix="fa"
                focusable="false"
                role="img"
                style={
                  Object {
                    "color": "#000",
                  }
                }
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </button>
        </div>
      </div>
    </div>
  </ExpertColorDetails>
</Provider>
`;
